x-pack/plugins/fleet/server/services/epm/elasticsearch/transform/install.ts

```diff
@@ -117,6 +134,218 @@ export const installTransform = async (
+
+  let previousInstalledTransformEsAssets: EsAssetReference[] = [];
+  if (installation) {
+    previousInstalledTransformEsAssets = installation.installed_es.filter(
+      ({ type, id }) => type === ElasticsearchAssetType.transform
+    );
+    if (previousInstalledTransformEsAssets.length) {
+      logger.info(
+        `Found previous transform references:\n ${JSON.stringify(
+          previousInstalledTransformEsAssets
+        )}`
+      );
+    }
+  }
+
+  // delete all previous transform
+  await deleteTransforms(
+    esClient,
+    previousInstalledTransformEsAssets.map((asset) => asset.id)
+  );
+
+  const installNameSuffix = `${installablePackage.version}`;
+  let installedTransforms: EsAssetReference[] = [];
+  if (transformPaths.length > 0) {
+    const transformsSpecifications = new Map();
+    const destinationIndexTemplates: DestinationIndexTemplateInstallation[] = [];
+    const indices = [];
+    const transforms: TransformInstallation[] = [];
+
+    transformPaths.forEach((path: string) => {
+      const { transformModuleId, fileName } = getTransformFolderAndFileNames(
+        installablePackage,
+        path
+      );
+
+      // Since there can be multiple assets per transform definition
+      // We want to create a unique list of assets/specifications for each transform
+      if (transformsSpecifications.get(transformModuleId) === undefined) {
+        transformsSpecifications.set(transformModuleId, new Map());
+      }
+      const packageAssets = transformsSpecifications.get(transformModuleId);
+
+      const content = safeLoad(getAsset(path).toString('utf-8'));
+
+      if (fileName === 'fields') {
+        const validFields = processFields(content);
+        const mappings = generateMappings(validFields);
+        packageAssets?.set('mappings', mappings);
+      }
+
+      if (fileName === 'transform') {
+        transformsSpecifications.get(transformModuleId)?.set('destinationIndex', content.dest);
+        indices.push(content.dest);
+        transformsSpecifications.get(transformModuleId)?.set('transform', content);
+        content._meta = getESAssetMetadata({ packageName: installablePackage.name });
+        transforms.push({
+          transformModuleId,
+          installationName: getTransformNameForInstallation(
+            installablePackage,
+            transformModuleId,
+            installNameSuffix
+          ),
+          content,
+        });
+        indices.push(content.dest);
+      }
+
+      if (fileName === 'manifest') {
+        if (isPopulatedObject(content, ['start']) && content.start === false) {
+          transformsSpecifications.get(transformModuleId)?.set('start', false);
+        }
+        // If manifest.yml contains destination_index_template
+        // Combine the mappings and other index template settings from manifest.yml into a single index template
+        // Create the index template and track the template in EsAssetReferences
```

Comment:
This is interesting information, but also implies immense complexity, so I think we'll need to take a step back and set some boundaries on the scope of transforms in packages.

The complexity comes because if the package data stream, which provides the transform's source data, can be infinitely customised by the user then the transform itself may need to be customised to match.

In the case of a `pivot` transform maybe it's OK to say that the transform doesn't do anything with the custom fields, so the transform destination index wouldn't need to contain them.

For a `latest` transform maybe we can just include the same `@custom` component template in the destination mappings that was included in the source mappings, and make a rule that the fields used to determine "latest" have to be defined by the package and not be custom fields.

This might avoid the need to have the ability to change the behaviour of the transform itself.

It's also interesting that in the transforms that have been added to packages in the past the namespace has been hardcoded as `default`, so there hasn't been any attempt in the past to make a packaged transform generic enough to work with different namespaces with different mappings - see for example: https://github.com/elastic/elasticsearch/blob/498a24b61f5e37148f48e16c9d47d6c49af08aac/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java#L808-L815

What's the package with the most complex combination of component templates we have today?  I think it would help if we could look at how all the flexibility is currently being used.

Reply:
> What's the package with the most complex combination of component templates we have today? I think it would help if we could look at how all the flexibility is currently being used.

AWS ships probably the most component template of any integration since it has so many policy templates. 

![image](https://user-images.githubusercontent.com/6766512/190002599-03a31d04-e617-4b1b-9af6-b9b09ee01f32.png)

If you install the AWS integration w/ many of its inputs enabled you'll get quite a lot of component templates installed. I don't know if this is necessarily "complex" as they don't do anything too involved from what I can tell, but maybe this is helpful here.

Reply:
We haven't discussed customization as a requirement for transforms in packages and IMO we should exclude that from the scope for now. I do think it's a good idea to avoid any manual merging on the Kibana side if possible.

I'm curious if we actually need to support both `destination_index_template` and `mappings` in the package-spec? Would it be possible to only allow one or the other and throw an error otherwise?

Reply:
After some discussion with Mark, I [switched to using buildComponentTemplates](https://github.com/elastic/kibana/pull/140046/commits/5a98445cfc4c630186c503ad2cfca7893628b6a8) helper to handle the logic for better consistency with what Fleet is doing. However, if we think it's better to throw error instead of both are defined I can definitely do that as well.